unified_planning/model/expression.py

```diff
@@ -92,7 +92,7 @@ def create_node(self, node_type: OperatorKind, args: Iterable['up.model.fnode.FN
         if content in self.expressions:
             return self.expressions[content]
         else:
-            n =up.model.fnode.FNode(content, self._next_free_id)
+            n = up.model.fnode.FNode(content, self._next_free_id, self.env)
```

Comment:
At this point, I would assert/check-and-except that all the `args` have the same environment and that it is `self.env`